[Projection Support] Addressing some feedback from PR #38

dnachev opened the pull request:
This addresses some of the feedback from PR #38 

- Grouped the projection related properties in the `init` function.
- Converted `_internalModel` to be a getter, which always delegates to the `baseModel`.
- Cleaned up the `merge` function and removed the check for `Date` in the `isObject` function.
- Removed the `modelIsProjection` function from the `SchemaManager`

runspired commented:
@dnachev does this need to point at the `cleanup-notifications` branch or can it be safely pointed straight into the m3-projections branch instead?

dnachev commented:
@runspired Apparently, simple rebase is not enough. I haven't looked deeper.

runspired commented on addon/model.js:
```diff
@@ -24,7 +24,7 @@ const FakeAttributeMeta = { isAttribute: true };
   constructor(record) {
     this.record = record;
-    this.modelName = this.record._internalModel.modelName;
+    this.modelName = this.record.modelName;
```
shouldn't this either be `this.modelName = this.record._modelName;` or M3._modelName be renamed to M3.modelName ?

dnachev replied:
Good catch! Thinking about it, the change was supposed to not use `_internalModel`, because it will refer to the base model, but `EmbeddedSnapshot` is only used for nested models, which doesn't have `_baseModel`, therefore the line should work as it is. I will revert it.

hjdivad replied:
should add a test for this

dnachev commented:
@runspired Addressed your comment.

runspired commented:
@rwjblue you have any ideas why we're seeing these unexpected browser crashes?

runspired commented on addon/util.js:
```diff
@@ -26,7 +26,7 @@ export const OWNER_KEY = (function() {
 })();
 
 export function isObject(value) {
```
I'm not sure I like the name `isObject` for this util anymore, maybe `isExtantNonArrayObject` ?

dnachev replied:
I had to see what "Extant" means :)

If you elaborate what is your concern of this being named `isObject`, we may be able to find better name.

runspired replied:
Extant basically just means "exists" or in this case "not null". My main thing is that `isObject` usually does not also check `!isArray` so I'd like to make that clear. And I figured if we are going to go so far as `isNonArrayObject` I might as well slip in a `not array and not null` point to it.

runspired replied:
I'd also be happy with `isNonArrayObject` though.

hjdivad replied:
`Extant` is not quite right.  It has a strong connotation of "still around" which is not really what we're testing here.

`isObject` seems ok if we change the implementation to something like `value && value.constructor === Object`

Basically we're looking for POJOs.

dnachev replied:
I like changing it to `value.constructor === Object`. Any objections? @runspired @hjdivad

rwjblue replied:
Checking constructor prevents using Object.create(null) objects, will that be an issue for us?

Do we care about anything other than `typeof obj === ‘object’ && obj !== null`?

runspired replied:
We care that it's not an array. David and I discussed this OOB this evening and think we should just rename the existing implantation (not null not array type object) to something like "isEmbeddedObject" or some such

runspired replied:
Basically rename to the reason for the checks instead of what the check is

dnachev replied:
OK, renamed the function to `isEmbeddedObject`, because:
- embedded is already the term we use to refer to the sub-models, which can be part of a model/record
- we use the function in two different places (not in the `merge` function) and although they are similar uses (both determining whether the logic should recurse), they are different to make it difficult to find a name, which fits both cases.

runspired commented on addon/model.js:
```diff
@@ -290,7 +290,15 @@ export default class MegamorphicModel extends Ember.Object {
 
   get _modelName() {
-    return this._internalModel.modelName;
+    return this.__internalModel.modelName;
```
why wouldn't this go through the `_internalModel` codepath ?

runspired commented on addon/model.js:
```diff
@@ -290,7 +290,15 @@ export default class MegamorphicModel extends Ember.Object {
+  }
+
+  set _internalModel(internalModel) {
+    this.__internalModel = internalModel;
```
Are we sure `get _internalModel() {}` should prefer `baseModel._internalModel` when present instead of `__internalModel` ? otherwise it seems we risk trolling ourselves by setting an _internalModel that we don't access back.

runspired commented:
It appears the test crashes are a known problem and Travis has reverted a change on their end. If the issue persists much longer we can configure chrome via `testem.js` to use `--no-sandbox` or set `sudo: required` in `.travis.yml`.

https://github.com/travis-ci/travis-ci/issues/8836

dnachev commented:
I've replaced the `_internalModel` getter to be `_data` as we actually care about the base model's `_data` and not the actual `_internalModel`. This also removed the couple of questionable references and ambiguities, which @runspired pointed out.
